rotkehlchen/rotkehlchen.py

```diff
@@ -106,6 +114,20 @@
 ICONS_QUERY_SLEEP = 60
 
 
+def kill_greenlet_and_acquire_locks(
+        stack: contextlib.ExitStack,
+        acquired_by_this_method: dict[Semaphore, bool],
+        greenlet: gevent.Greenlet,
+        locks: list[Semaphore],
+) -> None:
+    """Kills the provided greenlet and tries to acquire the locks."""
+    greenlet.kill(exception=GreenletKilledError('Killed to delete evm addresses'))
+    for lock in locks:
+        if lock.locked() is False:
```

Comment:
When a greenlet is killed, it is not guaranteed that the lock is released. For example, imagine this situation:
1. A periodic task for querying transactions runs (and it acquires a lock)
2. A request is made from frontend that also requests transactions (and it lines up in the queue for that lock)

Then we start killing greenlets. First we kill the api task. But since the lock is acquired by the periodic task, it does not free the lock, so we can't be sure here that the lock is acquirable.

Reply:
Yes but this still does not make sense to me.

> Then we start killing greenlets

The moment you kill the greenlet, you should get the locks it has acquired. No need to complicate things. Since `greenlet.kill()` kills it asynchronously what you would need todo I guess here is kill and wait for it.

And then yet again I don't understand why not just do a blocking acquire at the very end and be done with it? Since this is what you end up doing anyway. We can take the conversation to discord also.